kevm-pyk/src/kevm_pyk/utils.py

```diff
@@ -1,26 +1,113 @@
+    max_depth: int = 100,
+    max_iterations: Optional[int] = None,
+    workers: int = 1,
+    break_every_step: bool = False,
+    break_on_calls: bool = False,
+    implication_every_block: bool = False,
+    rpc_base_port: Optional[int] = None,
+    is_terminal: Optional[Callable[[CTerm], bool]] = None,
+    extract_branches: Optional[Callable[[CTerm], Iterable[KInner]]] = None,
+) -> int:
+    def _call_rpc(packed_args: Tuple[str, KCFG, int]) -> bool:
+        _cfgid, _cfg, _index = packed_args
+        terminal_rules = ['EVM.halt']
+        if break_every_step:
+            terminal_rules.append('EVM.step')
+        if break_on_calls:
+            terminal_rules.extend(
+                [
+                    'EVM.call',
+                    'EVM.callcode',
+                    'EVM.delegatecall',
+                    'EVM.staticcall',
+                    'EVM.create',
+                    'EVM.create2',
+                    'FOUNDRY.foundry.call',
+                ]
+            )
+        base_port = rpc_base_port if rpc_base_port is not None else find_free_port()
+        cfg_path = None
+        if save_directory is not None:
+            if _cfgid.isalnum():
+                _cfg_path = _cfgid
+            else:
+                hash = hashlib.sha256()
+                hash.update(_cfgid.encode('utf-8'))
+                _cfg_path = str(hash.hexdigest())
+                _LOGGER.info(f'Using hashed path for storing KCFG: {_cfgid} -> {_cfg_path}')
+            cfg_path = save_directory / f'{_cfg_path}.json'
+
+        with KCFGExplore(kprove, port=(base_port + _index)) as kcfg_explore:
+            try:
+                _cfg = kcfg_explore.all_path_reachability_prove(
+                    _cfgid,
+                    _cfg,
+                    cfg_path=cfg_path,
+                    is_terminal=is_terminal,
+                    extract_branches=extract_branches,
+                    max_iterations=max_iterations,
+                    execute_depth=max_depth,
+                    terminal_rules=terminal_rules,
+                    implication_every_block=implication_every_block,
+                )
+            except Exception as e:
+                _LOGGER.error(f'Proof crashed: {_cfgid}\n{e}')
+                return False
+
+        failure_nodes = _cfg.frontier + _cfg.stuck
+        if len(failure_nodes) == 0:
+            _LOGGER.info(f'Proof passed: {_cfgid}')
+            return True
+        else:
+            _LOGGER.error(f'Proof failed: {_cfgid}')
+            return False
+
+    with ProcessPool(ncpus=workers) as process_pool:
+        _proof_problems = [(_id, _cfg, _i) for _i, (_id, _cfg) in enumerate(proof_problems.items())]
+        results = process_pool.map(_call_rpc, _proof_problems)
+
+    failed = 0
+    for cid, succeeded in zip(proof_problems.keys(), results, strict=True):
+        if succeeded:
+            print(f'PASSED: {cid}')
+        else:
+            print(f'FAILED: {cid}')
```

Comment:
Consider pulling this side effect to the caller.

Reply:
Good idea, done!